### PR TITLE
Paint chairlift

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

**Comment by marijnvdwerf** [comment]
@Gymnasiast Can you remove the `work in progress` tag and add `paint-setup` + `pending review`?

File: src/ride/transport/chairlift.c [hunk 1]
```diff
@@ -13,3 +13,796 @@
...
+#include "../../world/map.h"
+#include "../track_paint.h"
+#include "../track.h"
+#include "../../paint/paint.h"
+#include "../../interface/viewport.h"
+#include "../../paint/supports.h"
+#include "../ride.h"
+#include "../ride_data.h"
+
+enum
```
Comment (marijnvdwerf) [inline src/ride/transport/chairlift.c]: These are used for multiple rides, so they should probably be moved

File: src/ride/transport/chairlift.c [hunk 2]
```diff
@@ -13,3 +13,796 @@
...
+static void chairlift_paint_station(uint8 rideIndex, uint8 trackSequence, uint8 direction, int height, rct_map_element * mapElement)
+{
+    if (direction % 2) {
+        chairlift_paint_station_se_nw(rideIndex, trackSequence, direction, height, mapElement);
+    } else {
+        chairlift_paint_station_ne_sw(rideIndex, trackSequence, direction, height, mapElement);
+    }
+}
+
+static void chairlift_paint_flat(uint8 rideIndex, uint8 trackSequence, uint8 direction, int height, rct_map_element * mapElement)
```
Comment (marijnvdwerf) [inline src/ride/transport/chairlift.c]: Missing address comment

File: src/ride/transport/chairlift.c [hunk 3]
```diff
@@ -13,3 +13,796 @@
...
+    SEGMENT_6 = SEGMENT_CC,
+    SEGMENT_D0 = (1 << 7),
+    SEGMENT_7 = SEGMENT_D0,
+    SEGMENT_D4 = (1 << 8),
+    SEGMENT_8 = SEGMENT_D4,
+};
+
+const int SEGMENTS_ALL = SEGMENT_B4 | SEGMENT_B8 | SEGMENT_BC | SEGMENT_C0 | SEGMENT_C4 | SEGMENT_C8 | SEGMENT_CC | SEGMENT_D0 | SEGMENT_D4;
+
+static void paint_util_set_segment_support_height(int flags, uint16 height, uint8 segment_flags)
```
Comment (marijnvdwerf) [inline src/ride/transport/chairlift.c]: We'll also need a segment rotation method, for tiles with a smaller footprint. Probably easiest to put the edge segments in an `uint8` and `ror8` them by `2 * rotation`. Not needed for this ride though, so not implemented yet

File: src/ride/transport/chairlift.c [hunk 4]
```diff
@@ -13,3 +13,796 @@
...
+        sub_98197C(imageId, 0, 16, 1, 1, 7, height + 2, 1, 16, height + 2, get_current_rotation());
+    }
+
+    if (drawFrontColumn) {
+        imageId = SPR_CHAIRLIFT_STATION_COLUMN_NE_SW | RCT2_GLOBAL(0x00F44198, uint32);
+        sub_98197C(imageId, 30, 16, 1, 1, 7, height + 2, 1, 16, height + 2, get_current_rotation()); // bound offset x is wrong?
+    }
+
+    paint_util_set_segment_support_height(SEGMENTS_ALL, 0xFFFF, 0);
+    paint_util_push_tunnel_left(height, TUNNEL_FLAT);
```
Comment (marijnvdwerf) [inline src/ride/transport/chairlift.c]: Tunnel-pushing for end stations should probably get checked.

**Comment by marijnvdwerf** [comment]
image

Any ideas on naming the sprites would be welcome. Any winter sports enthousiasts?

**Comment by Gymnasiast** [comment]
The large wheel is called the 'bullwheel'.
For the other pieces I'd just take the track piece descriptors as inspiration:
`SKILIFT_ELEM_25_DEG_DOWN_TO_FLAT_NESW` and so on.

File: src/ride/transport/chairlift.c [hunk 5]
```diff
@@ -13,3 +13,796 @@
...
+        if (flags & (1 << s)) {
+            RCT2_GLOBAL(0x0141E9B4 + s * 4, uint16) = height;
+            RCT2_GLOBAL(0x0141E9B6 + s * 4, uint8) = segment_flags;
+        }
+    }
+}
+
+static void paint_util_set_support_height(uint16 height, uint8 flags)
+{
+    if (RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT, sint16) < height) {
```
Comment (janisozaur) [inline src/ride/transport/chairlift.c]: shouldn't `RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT` be accessed as `uint16`?

File: src/ride/transport/chairlift.c [hunk 6]
```diff
@@ -13,3 +13,796 @@
...
+        case EDGE_NW:
+            offset = loc_7667AE[rotation];
+            break;
+    };
+
+    uint16 entranceLoc =
+        ((position.x / 32) + offset.x) |
+        (((position.y / 32) + offset.y) << 8);
+
+    uint8 entranceId = (mapElement->properties.track.sequence & 0x70) >> 4;
```
Comment (janisozaur) [inline src/ride/transport/chairlift.c]: you should be using
```
src/world/map.c:3768: int map_get_station(rct_map_element *mapElement)
```

File: src/ride/transport/chairlift.c [hunk 7]
```diff
@@ -13,3 +13,796 @@
...
+        uint16 temp = RCT2_GLOBAL(0x0141E9B4 + s * 4, uint16);
+        RCT2_GLOBAL(0x0141E9B4 + s * 4, uint16) = RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_PAINT_TILE_MAX_HEIGHT, uint16);
+        metal_a_supports_paint_setup(10, s, 0, height, RCT2_GLOBAL(0x00F4419C, uint32));
+        RCT2_GLOBAL(0x0141E9B4 + s * 4, uint16) = temp;
+    }
+}
+
+rct_map_element * map_get_track_element_at_from_ride_fuzzy(int x, int y, int z, int rideIndex)
+{
+    rct_map_element * mapElement = map_get_first_element_at(x >> 5, y >> 5);
```
Comment (janisozaur) [inline src/ride/transport/chairlift.c]: check for `mapElement == NULL`

**Pull request merged**